index/segment/types.go

```diff
@@ -44,66 +35,35 @@ type Segment interface {
 
 	// Size returns the number of documents indexed.
 	Size() uint32
+}
 
-	// Iter returns a segment document iterator.
-	Iter() Iter
+// Readable is a readable index segment.
+type Readable interface {
+	// Query retrieves the list of documents matching the given criterion.
+	Query(q Query) (ResultsIter, error)
 }
 
-// Iter allows for iteration over the documents indexed in the segment.
-type Iter interface {
+// ResultsIter is an iterator over query results.
+type ResultsIter interface {
 	// Next returns a bool indicating if the iterator has any more documents
 	// to return.
 	Next() bool
 
-	// Current returns the current document and the DocID associated with the
-	// document in the segment.
-	Current() (doc.Document, DocID)
-
-	// Err returns any errors the Iter ran into whilst iterating.
-	Err() error
-}
-
-// Readable is a readable index segment.
-type Readable interface {
-	// Query retrieves the list of documents matching the given criterion.
-	Query(q Query) []doc.Document
+	// Current returns the current document and whether the document was marked
+	// as deleted or not.
+	Current() (d doc.Document, tombstoned bool)
 }
 
 // Writable is a writable index segment.
 type Writable interface {
 	// Insert inserts the given documents with provided fields.
 	Insert(d doc.Document) error
 
-	// Update updates the given document.
-	Update(d doc.Document) error
-
 	// Delete deletes the given document.
 	Delete(d doc.Document) error
 }
 
-// Writer represents a segment writer.
-type Writer interface {
-	Writable
-
-	// Open prepares the writer to accept writes.
-	Open() error
-
-	// Close closes the writer.
-	Close() error
-}
-
-// Reader represents a segment reader.
-type Reader interface {
-	Readable
-
-	// Open prepares the reader to accept reads.
-	Open() error
-
-	// Close closes the reader.
-	Close() error
-}
-
-// DocID is document identifier internal to each index segment. It's limited to
+// DocID is a document identifier internal to each index segment. It's limited to
 // be at most 2^32.
 type DocID uint32
```

Comment:
Any reason not to use int64?  4bill seems too small, especially since we already have 4bill in our current index (granted that is globally not per zone).

Reply:
We picked `int32` because there's much greater support for it in algorithms for bitmap indices (Roaring) and posting lists compression (Frame of Reference and Simple Family) than `int64`. Not that they couldn't be adopted for `int64` just that it would require more up-front work on our end (for example, we can use the go implementation of Roaring as is). We still haven't settled on this though. I think when we last left off I was in favor of an ID per index, with `int64` if we felt we would have indices with more than 4 billion documents, and Prateek was in favor of an ID per segment, with `int32` since it's how to imagine a single segment containing more than 4 billion IDs, and we were going to revisit this issue later on when we had more benchmarks. Which is all to say I think this is still largely up in the air lol

Reply:
Pretty much what Jerome said. Only thing to add is the DocID is internal to each Segment, which are per node. I don't think we'll ever hit a single node with 4B documents so this isn't a size restriction.